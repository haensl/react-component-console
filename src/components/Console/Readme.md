A console emulator component.

Example:

```js
<Console lines={[
  'Hi!',
  'How are you today?',
  'Whatup?'
]}/>
```
