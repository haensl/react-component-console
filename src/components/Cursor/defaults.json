{
  "classes": {
    "blink": "Console-Cursor--blink",
    "element": "Console-Cursor",
    "write": "Console-Cursor--write"
  },
  "intervalMs": 400
}
