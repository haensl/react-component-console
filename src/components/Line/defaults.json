{
  "classes": {
    "element": "Console-Line",
    "content": "Console-Line-content"
  }
}
