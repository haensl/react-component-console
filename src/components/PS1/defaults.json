{
  "classes": {
    "element": "Console-PS1"
  },
  "content": "$"
}
