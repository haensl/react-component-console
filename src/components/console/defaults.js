import { defaults as ps1Defaults } from '../ps1';
import { defaults as lineDefaults } from '../line';
import { defaults as cursorDefaults } from '../cursor';

export default {
  console: {
    append: false,
    classes: {
      element: 'console'
    },
    typing: {
      char: {
        avgMs: 150,
        deviation: 0.3,
        minMs: 50,
        maxMs: Infinity
      },
      line: {
        delay: {
          avgMs: 1000,
          deviation: 0.5,
          minMs: 500,
          maxMs: 1500
        }
      }
    }
  },
  line: lineDefaults,
  ps1: ps1Defaults,
  cursor: cursorDefaults
};
