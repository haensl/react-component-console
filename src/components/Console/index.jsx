import React from 'react';
import merge from 'deepmerge';
import regeneratorRuntime from 'regenerator-runtime';
import Line from '../Line';
import defaults from './defaults';
import math from '../../util/math';
import array from '../../util/array';

const isClient = typeof window === 'object';
const initialState = {
  console: {
    text: '',
    hasFinishedWritingLines: false
  },
  lines: null,
  initialized: false
};

const overwriteArrays = (dst, src, opts) => src;

class Console extends React.Component {
  constructor() {
    super();
    this.state = merge({}, initialState);
  }

  getDelay(opts = defaults.console.typing.char) {
    const sign = math.randSign();
    const deviation = Math.random() * opts.deviation;
    const delay = opts.avgMs
      + (opts.avgMs * deviation * sign);
    return Math.max(
      0,
      Math.min(
        opts.maxMs,
        Math.max(
          opts.minMs,
          delay
        )
      )
    );
  }

  async consumeLine(line) {
    if (line.length) {
      const c = line.slice(0, 1);
      await new Promise((resolve) => {
        this.setState({
          console: merge(
            this.state.console,
            {
              text: `${this.state.console.text}${c}`
            }
          ),
          cursor: merge(
            this.state.cursor,
            {
              char: c
            }
          ),
          timeouts: (this.state.timeouts || []).concat([
            window.setTimeout((() =>
              this.consumeLine(line.slice(1)).then(resolve))
                .bind(this),
              this.getDelay(this.state.console.typing.char) // eslint-disable-line
            )
          ])
        });
      });
    }

    return Promise.resolve();
  }

  async writeLine(line) {
    await this.awaitableSetState({
      console: merge(
        this.state.console,
        {
          text: ''
        }
      )
    });
    await this.consumeLine(line);
    await this.awaitableSetState({
      cursor: merge(
        this.state.cursor,
        {
          char: null
        }
      )
    });
    const self = this;
    return new Promise((resolve) => {
      self.setState({
        timeouts: (self.state.timeouts || []).concat([
          window.setTimeout(() => {
            resolve();
          }, self.getDelay(self.state.console.typing.line.delay))
        ])
      });
    });
  }

  awaitableSetState(state) {
    const self = this;
    return new Promise((resolve) => {
      self.setState(state, resolve);
    });
  }

  async *linesGenerator() {
    if (!(Array.isArray(this.state.lines) && this.state.lines.length)) {
      return;
    }

    for (let currentLine = this.state.console.currentLine || 0; currentLine < this.state.lines.length; currentLine++) {
      await this.awaitableSetState({
        console: merge(
          this.state.console,
          {
            currentLine
          }
        )
      });

      await this.writeLine(this.state.lines[currentLine]);

      if (this.state.console.append) {
        await this.awaitableSetState({
          console: merge(
            this.state.console,
            {
              lines: [this.state.lines[currentLine]]
            }
          )
        });
      }

      yield this.state.lines[currentLine];
    }
  }

  async writeLines() {
    await this.awaitableSetState({
      console: merge(
        this.state.console,
        { hasFinishedWritingLines: false }
      )
    });

    for await (const line of this.linesGenerator()) {
      if (typeof this.props.onFinishWritingLine === 'function') {
        this.props.onFinishWritingLine(line);
      }
    }
    const newState = merge({}, this.state);
    newState.console.hasFinishedWritingLines = true;

    if (this.state.console.append) {
      newState.console.text = '';
    }

    this.clearTimeouts();
    await this.awaitableSetState(newState);
    if (typeof this.props.onFinishWritingLines === 'function') {
      this.props.onFinishWritingLines(this.state.lines);
    }
  }

  get isWriting() {
    return typeof this.state.console.currentLine === 'number'
      && !this.state.console.hasFinishedWritingLines;
  }

  async initialize(props) {
    if (isClient) {
      await this.updateFromProps(
        props,
        {
          ...initialState,
          lines: this.state.lines,
          initialized: true
        }
      );

      const lines = this.linesFromProps(props);
      if (lines) {
        this.onReceiveLines(lines);
      }
    }
  }

  linesFromProps(props) {
    if (!props || !props.lines) {
      return null;
    }

    if (Array.isArray(props.lines)) {
      return props.lines.map((line) => line.toString());
    } else {
      return [ props.lines.toString() ];
    }
  }

  async updateFromProps(props, overwrites = {}) {
    await this.awaitableSetState(
      merge.all([
        defaults,
        this.state,
        props,
        overwrites
      ], {
        arrayMerge: overwriteArrays
      })
    );
  }

  async onReceiveLines(nextLines) {
    if (!Array.isArray(nextLines)) {
      throw new Error(`Invalid parameter: Lines. Expected Array. Received ${typeof nextLines}\n${nextLines}`);
    }

    if (!this.state.lines) {
      await this.awaitableSetState({
        console: merge(
          this.state.console,
          {
            text: '',
            currentLine: 0,
            hasFinishedWritingLines: false
          }
        ),
        lines: nextLines
      });
      this.writeLines();
      return;
    }

    if (array.equals(nextLines, this.state.lines)) {
      return;
    }

    // lines added
    if (nextLines.length > this.state.lines.length
      && array.isSubset(this.state.lines, nextLines, 0)) {
      await this.awaitableSetState({
        lines: nextLines
      });
      if (!this.isWriting || this.state.console.hasFinishedWritingLines) {
        await this.awaitableSetState({
          console: merge(
            this.state.console,
            {
              text: '',
              currentLine: this.state.console.currentLine + 1,
              hasFinishedWritingLines: false
            }
          )
        });
        this.writeLines();
      }
    // lines removed
    } else if (nextLines.length < this.state.lines.length
      && array.isSubset(nextLines, this.state.lines)) {
      await this.awaitableSetState({
        lines: nextLines
      });
      if (this.isWriting
        && this.state.console.currentLine >= nextLines.length) {
        await this.stopWriting();
        await this.awaitableSetState({
          console: merge(
            this.state.console,
            {
              text: nextLines[nextLines.length - 1],
              currentLine: nextLines.length - 1,
              hasFinishedWritingLines: true
            }
          )
        });
      } else if (!this.isWriting) {
        await this.awaitableSetState({
          console: merge(
            this.state.console,
            {
              text: nextLines[nextLines.length - 1],
              currentLine: nextLines.length - 1,
              hasFinishedWritingLines: true
            }
          )
        });
      }
    // lines changed
    } else {
      if (this.isWriting) {
        this.stopWriting();
      }

      await this.awaitableSetState({
        console: merge(
          this.state.console,
          {
            text: '',
            currentLine: 0,
            hasFinishedWritingLines: false
          }
        ),
        lines: nextLines
      });
      this.writeLines();
    }
  }

  componentDidMount() {
    this.initialize(this.props);
  }

  componentWillReceiveProps(nextProps, nextState) {
    if (isClient && this.state.initialized) {
      const currentLines = Array.isArray(this.state.lines)
        ? [].concat(this.state.lines)
        : null;
      this.updateFromProps(nextProps, {
        lines: currentLines
      });
      const lines = this.linesFromProps(nextProps);
      if (lines) {
        this.onReceiveLines(lines);
      }
    }
  }

  componentWillUnmount() {
    this.stopWriting();
  }

  async clearTimeouts() {
    (this.state.timeouts || []).forEach((t) => window.clearTimeout(t));
    await this.awaitableSetState({
      timeouts: []
    });
  }

  async stopWriting() {
    if (isClient && this.state.timeouts) {
      await this.clearTimeouts();
      await this.awaitableSetState({
        console: merge(
          this.state.console,
          {
            hasFinishedWritingLines: false
          }
        )
      });
    }
  }

  render() {
    if (!this.state.initialized) {
      return null;
    }

    return (
      <div className={ this.state.console.classes.element }>
        {
          (this.state.console.lines || []).map((line) => (
            <Line classes={ this.state.line.classes } content={ line } />
          ))
        }
        <Line classes={ this.state.line.classes }
          write={ true }
          ps1={
            this.state.ps1
          }
          content={
            this.state.console.text
          }
          cursor={
            this.state.cursor
          }
        />
      </div>
    );
  }
}

export default Console;
