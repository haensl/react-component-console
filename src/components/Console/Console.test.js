import React from 'react';
import { shallow, mount } from 'enzyme';
import Console from './';
import defaults from './defaults';

jest.useFakeTimers();

const maxCharTimeout = defaults.console.typing.char.avgMs
  + (defaults.console.typing.char.deviation * defaults.console.typing.char.avgMs);
const maxNewlineTimeout = defaults.console.typing.line.delay.avgMs
  + (defaults.console.typing.line.delay.deviation * defaults.console.typing.line.delay.avgMs);
const waitForAsync = () => new Promise((resolve) => setImmediate(resolve));

describe('Console', () => {
  let tree;
  let component;

  describe('empty', () => {
    beforeEach(() => {
      tree = shallow(<Console />);
    });

    it('renders as expected', () => {
      expect(tree).toMatchSnapshot();
    });

    it('adds the default class to the element', () => {
      expect(tree.hasClass(defaults.console.classes.element)).toBe(true);
    });

    it('renders the Line component', () => {
      expect(tree.find('Line').length).toBe(1);
    });
  });

  describe('writing', () => {
    describe('string', () => {
      let line;
      let lineCallback;
      let linesCallback;

      beforeEach(async () => {
        line = 'test line.';
        lineCallback = jest.fn();
        linesCallback = jest.fn();
        component = mount(
          <Console
            lines={ line }
            onFinishWritingLines={ linesCallback }
            onFinishWritingLine={ lineCallback }
          />
        );
        await waitForAsync();
      });

      it('renders as expected', () => {
        component.update();
        expect(component).toMatchSnapshot();
      });

      describe('Cursor', () => {
        it('writes the first character', () => {
          component.update();
          expect(component.find('Cursor').prop('char')).toEqual('t');
        });
      });

      describe('Line', () => {
        it('writes the first character', () => {
          component.update();
          expect(component.find('Line').prop('content')).toEqual('t');
        });
      });

      describe('after some time', () => {
        beforeEach(() => {
          jest.advanceTimersByTime(maxCharTimeout);
          component.update();
        });

        it('renders as expected', () => {
          expect(component).toMatchSnapshot();
        });

        describe('Cursor', () => {
          it('writes the second character', () => {
            expect(component.find('Cursor').prop('char')).toEqual('e');
          });
        });

        describe('Line', () => {
          it('writes the first two characters', () => {
            expect(component.find('Line').prop('content')).toEqual('te');
          });
        });

        describe('after the whole line has been written', () => {
          beforeEach(() => {
            for (let i = 0; i < line.length; i++) {
              jest.advanceTimersByTime(maxCharTimeout);
              component.update();
            }
          });

          it('renders as expected', () => {
            expect(component).toMatchSnapshot();
          });

          describe('Cursor', () => {
            it('writes the last character', () => {
              expect(component.find('Cursor').prop('char')).toEqual('.');
            });
          });

          describe('Line', () => {
            it('contains the whole line', () => {
              expect(component.find('Line').prop('content')).toEqual(line);
            });
          });

          describe('onFinishWritingLines', () => {
            beforeEach(async () => {
              await waitForAsync();
              jest.runOnlyPendingTimers();
              await waitForAsync();
              jest.runOnlyPendingTimers();
            });

            it('calls the onFinishWritingLines callback', () => {
              expect(linesCallback).toHaveBeenCalled();
            });
          });

          describe('onFinishWritingLine', () => {
            beforeEach(async () => {
              await waitForAsync();
              jest.runOnlyPendingTimers();
            });

            it('calls the onFinishWritingLines callback', () => {
              expect(lineCallback).toHaveBeenCalledWith(line);
            });
          });
        });
      });
    });

    describe('array of strings', () => {
      let line;
      let lines;
      let props;
      let lineCallback;
      let linesCallback;

      beforeEach(async () => {
        lineCallback = jest.fn();
        linesCallback = jest.fn();
        lines = [
          'test line.',
          'another line.'
        ];
        props = {
          lines
        };
        component = mount(
          <Console { ...props }
            onFinishWritingLine={ lineCallback }
            onFinishWritingLines={ linesCallback }
          />
        );
        await waitForAsync();
      });

      describe('first line', () => {
        beforeEach(() => {
          line = lines[0];
          component.update();
        });

        it('renders as expected', () => {
          expect(component).toMatchSnapshot();
        });

        describe('Cursor', () => {
          it('writes the first character', () => {
            expect(component.find('Cursor').prop('char')).toEqual(line.slice(0, 1));
          });
        });

        describe('Line', () => {
          it('writes the first character', () => {
            expect(component.find('Line').prop('content')).toEqual(line.slice(0, 1));
          });
        });

        describe('after some time', () => {
          beforeEach(() => {
            jest.advanceTimersByTime(maxCharTimeout);
            component.update();
          });

          it('renders as expected', () => {
            expect(component).toMatchSnapshot();
          });

          describe('Cursor', () => {
            it('writes the second character', () => {
              expect(component.find('Cursor').prop('char')).toEqual(line.slice(1, 2));
            });
          });

          describe('Line', () => {
            it('writes the first two characters', () => {
              expect(component.find('Line').prop('content')).toEqual(line.slice(0, 2));
            });
          });

          describe('after the whole line has been written', () => {
            beforeEach(() => {
              for (let i = 2; i < line.length; i++) {
                jest.advanceTimersByTime(maxCharTimeout);
                component.update();
              }
            });

            it('renders as expected', () => {
              expect(component).toMatchSnapshot();
            });

            describe('Cursor', () => {
              it('writes the last character', () => {
                expect(component.find('Cursor').prop('char')).toEqual(line.slice(-1));
              });
            });

            describe('Line', () => {
              it('contains the whole line', () => {
                expect(component.find('Line').prop('content')).toEqual(line);
              });
            });

            describe('second line', () => {
              beforeEach(async () => {
                line = lines[1];
                await waitForAsync();
                jest.advanceTimersByTime(maxNewlineTimeout);
                await waitForAsync();
                component.update();
              });

              describe('onFinishWritingLine', () => {
                it('calls the onFinishWritingLine callback', () => {
                  expect(lineCallback).toHaveBeenCalledWith(lines[0]);
                });
              });

              it('renders as expected', () => {
                expect(component).toMatchSnapshot();
              });

              describe('Cursor', () => {
                it('writes the first character of the second line', () => {
                  expect(component.find('Cursor').prop('char')).toEqual(line.slice(0, 1));
                });
              });

              describe('Line', () => {
                it('contains the first character', () => {
                  expect(component.find('Line').prop('content')).toEqual(line.slice(0, 1));
                });
              });

              describe('after some time', () => {
                beforeEach(() => {
                  jest.advanceTimersByTime(maxCharTimeout);
                  component.update();
                });

                it('renders as expected', () => {
                  expect(component).toMatchSnapshot();
                });

                describe('Cursor', () => {
                  it('writes the second character of the second line', () => {
                    expect(component.find('Cursor').prop('char')).toEqual(line.slice(1, 2));
                  });
                });

                describe('Line', () => {
                  it('contains the first two characters of the second line', () => {
                    expect(component.find('Line').prop('content')).toEqual(line.slice(0, 2));
                  });
                });

                describe('after the whole second line has been written', () => {
                  beforeEach(() => {
                    for (let i = 2; i < line.length; i++) {
                      jest.advanceTimersByTime(maxCharTimeout);
                      component.update();
                    }
                  });

                  it('renders as expected', () => {
                    expect(component).toMatchSnapshot();
                  });

                  describe('Cursor', () => {
                    it('writes the last character of the second line', () => {
                      expect(component.find('Cursor').prop('char')).toEqual(line.slice(-1));
                    });
                  });

                  describe('Line', () => {
                    it('contains the whole second line', () => {
                      expect(component.find('Line').prop('content')).toEqual(line);
                    });
                  });

                  describe('onFinishWritingLine', () => {
                    beforeEach(async () => {
                      await waitForAsync();
                      jest.runOnlyPendingTimers();
                    });

                    it('calls the onFinishWritingLine callback the second time', () => {
                      expect(lineCallback).toHaveBeenCalledTimes(2);
                    });
                  });

                  describe('onFinishWritingLines', () => {
                    beforeEach(async () => {
                      await waitForAsync();
                      jest.runOnlyPendingTimers();
                      await waitForAsync();
                      jest.runOnlyPendingTimers();
                    });

                    it('calls the onFinishWritingLines callback', () => {
                      expect(linesCallback).toHaveBeenCalled();
                    });
                  });
                });
              });
            });
          });
        });
      });
    });
  });
});

