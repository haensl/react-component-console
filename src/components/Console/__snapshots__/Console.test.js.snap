// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Console empty renders as expected 1`] = `
<div
  className="Console"
>
  <Line
    classes={
      Object {
        "content": "Console-Line-content",
        "element": "Console-Line",
      }
    }
    content=""
    cursor={
      Object {
        "classes": Object {
          "blink": "Console-Cursor--blink",
          "element": "Console-Cursor",
          "write": "Console-Cursor--write",
        },
        "intervalMs": 400,
      }
    }
    ps1={
      Object {
        "classes": Object {
          "element": "Console-PS1",
        },
        "content": "$",
      }
    }
    write={true}
  />
</div>
`;

exports[`Console writing array of strings first line after some time after the whole line has been written renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="test line."
      cursor={
        Object {
          "char": ".",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          test line.
        </span>
        <Cursor
          char="."
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            .
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing array of strings first line after some time after the whole line has been written second line after some time after the whole second line has been written renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={
    [MockFunction] {
      "calls": Array [
        Array [
          "test line.",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="another line."
      cursor={
        Object {
          "char": ".",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          another line.
        </span>
        <Cursor
          char="."
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            .
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing array of strings first line after some time after the whole line has been written second line after some time renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={
    [MockFunction] {
      "calls": Array [
        Array [
          "test line.",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="an"
      cursor={
        Object {
          "char": "n",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          an
        </span>
        <Cursor
          char="n"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            n
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing array of strings first line after some time after the whole line has been written second line renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={
    [MockFunction] {
      "calls": Array [
        Array [
          "test line.",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="a"
      cursor={
        Object {
          "char": "a",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          a
        </span>
        <Cursor
          char="a"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            a
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing array of strings first line after some time renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="te"
      cursor={
        Object {
          "char": "e",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          te
        </span>
        <Cursor
          char="e"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            e
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing array of strings first line renders as expected 1`] = `
<Console
  lines={
    Array [
      "test line.",
      "another line.",
    ]
  }
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="t"
      cursor={
        Object {
          "char": "t",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          t
        </span>
        <Cursor
          char="t"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            t
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing string after some time after the whole line has been written renders as expected 1`] = `
<Console
  lines="test line."
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="test line."
      cursor={
        Object {
          "char": ".",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          test line.
        </span>
        <Cursor
          char="."
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            .
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing string after some time renders as expected 1`] = `
<Console
  lines="test line."
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="te"
      cursor={
        Object {
          "char": "e",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          te
        </span>
        <Cursor
          char="e"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            e
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;

exports[`Console writing string renders as expected 1`] = `
<Console
  lines="test line."
  onFinishWritingLine={[MockFunction]}
  onFinishWritingLines={[MockFunction]}
>
  <div
    className="Console"
  >
    <Line
      classes={
        Object {
          "content": "Console-Line-content",
          "element": "Console-Line",
        }
      }
      content="t"
      cursor={
        Object {
          "char": "t",
          "classes": Object {
            "blink": "Console-Cursor--blink",
            "element": "Console-Cursor",
            "write": "Console-Cursor--write",
          },
          "intervalMs": 400,
        }
      }
      ps1={
        Object {
          "classes": Object {
            "element": "Console-PS1",
          },
          "content": "$",
        }
      }
      write={true}
    >
      <pre
        className="Console-Line"
      >
        <PS1
          classes={
            Object {
              "element": "Console-PS1",
            }
          }
          content="$"
        >
          <span
            className="Console-PS1"
          >
            $
          </span>
        </PS1>
        <span
          className="Console-Line-content"
        >
          t
        </span>
        <Cursor
          char="t"
          classes={
            Object {
              "blink": "Console-Cursor--blink",
              "element": "Console-Cursor",
              "write": "Console-Cursor--write",
            }
          }
          intervalMs={400}
        >
          <span
            className="Console-Cursor Console-Cursor--write"
          >
            t
          </span>
        </Cursor>
      </pre>
    </Line>
  </div>
</Console>
`;
